pinot-core/src/main/java/com/linkedin/pinot/core/query/pruner/TimeSegmentPruner.java

```diff
@@ -33,11 +38,80 @@
 
   @Override
   public boolean prune(IndexSegment segment, BrokerRequest brokerRequest) {
-    Interval interval = segment.getSegmentMetadata().getTimeInterval();
-    if (interval != null && brokerRequest.getTimeInterval() != null && !new Interval(brokerRequest.getTimeInterval()).contains(interval)) {
-      return true;
+    SegmentMetadata metadata = segment.getSegmentMetadata();
+    String timeColumn = metadata.getTimeColumn();
+
+    if (timeColumn == null) {
+      return false;
+    }
+
+    TimeInterval segmentInterval = new TimeInterval(metadata.getStartTime(), metadata.getEndTime());
+    FilterQueryTree filterQueryTree = RequestUtils.generateFilterQueryTree(brokerRequest);
+    return (filterQueryTree != null && pruneSegment(filterQueryTree, segmentInterval, timeColumn));
+  }
+
+  /**
+   * Helper method to determine if a segment can be pruned based on segment's time metadata and
+   * the predicates on time column. It's algorithm is as follows:
+   *
+   * <ul>
+   *   <li> For leaf node: Returns true if there is a predicate on time and apply the predicate would result
+   *   in filtering out all docs of the segment, false otherwise. </li>
+   *   <li> For non-leaf AND node: True if any of its children returned true, false otherwise. </li>
+   *   <li> For non-leaf OR node: True if all its children returned true, false otherwise. </li>
+   * </ul>
+   *
+   * @param filterQueryTree Filter tree for the query.
+   * @param segmentInterval Time interval of the segment to be pruned.
+   * @return True if segment can be pruned out, false otherwise.
+   */
+  public static boolean pruneSegment(@Nonnull FilterQueryTree filterQueryTree, @Nonnull TimeInterval segmentInterval,
+      @Nonnull String timeColumn) {
+    List<FilterQueryTree> children = filterQueryTree.getChildren();
+
+    // Leaf Node
+    FilterOperator operator = filterQueryTree.getOperator();
```

Comment:
Let me flip it and ask, if it can be done with similar code in O(Tree size + # segments), what is a good reason to do it in O(# segments * Tree size) cost ?

I believe <1 microsecond is for 1 segment. For 500 segments of a table (we are in that ballpark with some use-cases) it's about 400-500 microseconds. If clients happen to construct bad query programmatically with nested OR/ANDs with multiple of those having time columns, Tree_size in above formula increases. Numbers are very small. Latency impact is ~5-10% for low latency use cases.